apollo-federation/src/sources/connect/json_selection/apply_to.rs

```diff
@@ -242,53 +240,79 @@ impl ApplyTo for NamedSelection {
     }
 }
 
-// $typenames is a special variable for referring to literal typenames. See
-// note in selection_set.rs for more detail.
-pub(super) const TYPENAMES: &str = "$typenames";
-
 impl ApplyTo for PathSelection {
     fn apply_to_path(
         &self,
         data: &JSON,
-        vars: &IndexMap<String, JSON>,
-        input_path: &mut Vec<JSON>,
+        vars: &VarsWithPathsMap,
+        input_path: &InputPath<JSON>,
         errors: &mut IndexSet<ApplyToError>,
     ) -> Option<JSON> {
-        if let JSON::Array(array) = data {
-            return self.apply_to_array(array, vars, input_path, errors);
-        }
-
-        match self {
-            Self::Var(var_name, tail) => {
-                if var_name == "$" {
-                    // Because $ refers to the current value, we keep using
-                    // input_path instead of creating a new var_path here.
-                    tail.apply_to_path(data, vars, input_path, errors)
-                } else if var_name == TYPENAMES {
-                    if let PathSelection::Key(Key::Field(ref name), _) = **tail {
-                        let var_data = json!({ name: name });
-                        let mut var_path = vec![json!(name)];
-                        tail.apply_to_path(&var_data, vars, &mut var_path, errors)
+        match &self.path {
+            // If this is a KeyPath, instead of using data as given, we need to
+            // evaluate the path starting from the current value of $. To
+            // evaluate the KeyPath against data, prefix it with @. This logic
+            // supports method chaining like obj->has('a')->and(obj->has('b')),
+            // where both obj references are interpreted as $.obj.
+            PathList::Key(key, tail) => {
+                if let Some((dollar_data, dollar_path)) = vars.get("$") {
+                    let input_path_with_key = dollar_path.append(key.to_json());
+                    if let Some(child) = dollar_data.get(key.as_string()) {
```

Comment:
```suggestion
                    if let Some(child) = dollar_data.get(key.as_str()) {
```

We don't need to clone the key again here, if we add this method to `Key`:

```rust
pub fn as_str(&self) -> &str {
        match self {
            Key::Field(name) => name.as_str(),
            Key::Quoted(name) => name.as_str(),
        }
    }
```

Reply:
https://github.com/apollographql/router/pull/5762/commits/489cf5575a63a18f93c79b0a96d7deb28eb4c651